### ISSUE TITLE
new device model: read/writeLater

### ISSUE THREAD

**Issue by warner** [Agoric/agoric-sdk#55]
We (me, dean, mark, chris) had a long discussion today about device drivers in SwingSet. Not sure we came to any clear conclusions, but here's what I remember:

We need three properties:

* safely manage cross-Realm calls: protect against leaking primal-realm objects into the kernel-realm, by coercing all arguments to primitive types, and catching/rewriting exceptions
* maintain transcript-based persistence of vats
* provide synchronous (at least) read access to endowments

We have three devices in mind: Mailbox (exists), Timer/Clock (#148), and something to help integrate the cosmos-sdk Bank module's account entries with an Issuer Vat's purses.

Dean's experience suggests that the React model is the best one to follow: handler functions get a snapshot of the current state, plus a function that lets them queue changes to be applied at the end of the operation (iff i doesn't get cancelled or aborted for some reason).

We defined three levels of turns:

* a "Turn" takes us from one empty Javascript stack to the next empty stack
* at a "Crank" boundary, both the javascript stack and the Vat's runnable-promise queue are empty. There may still be unresolved promises with .then callbacks queued, but there will not be any resolved promises with .then callbacks. This corresponds to a `kernel.step()`, and is achieved by using `setImmediate()` to wait on the IO queue, which is strictly lower-priority than the promise queue. Swingset Vats cannot share JS Promises or resolver functions (they can only share data and reference slots, through the kernel), therefore nothing in a Vat can get control until the kernel next invokes the `dispatch()` function, even if other Vats resolve their own promises (i.e. the object graphs of separate Vats are disjoint except for their link through the kernel syscall/dispatch, and those calls don't accept JS promises)
* the "Batch" contains as many Cranks as we can do before saving the state to durable storage. In a blockchain environment, each Batch goes into a separate block. We end the Batch when we finish the entire run-queue, or when we've done too much work (e.g. we've run up against the blocksize limit, or gas limit for a single block), or when we want to checkpoint our state so we don't lose too much progress if the host crashes.

We were converging on a device design that exposes two functions to the calling vat: `syscall.read(devnode, args) -> results`, and `syscall.writeLater(devnode, args) -> error`. The first causes the device to be invoked with `dispatch.read(devnode, args)` and should return synchronous data, but not modify any state. The second would cause two device invocations. The first is a `checkWrite` that should compare the `args` against a shadow state object that it builds up, to see if the writes would succeed later (e.g. does a balance transfer underflow, or does an argument refer to a missing timer object). `checkWrite` is allowed to return an error, but it is not allowed to commit any state changes to its endowments. If it returns success, the kernel pushes a call to `dispatch.write` to a special queue that does not run until the end of the current Crank (so it will be abandoned if any other Turn causes a vat-killing error). At the end of the Crank, all device writes are delivered, giving the device an opportunity to commit the changes to the endowments.

Rather than `dispatch`-style calls, another possible device API could be:

* `read(devnode, args) -> data`
* `write(devnode, args, S) -> error or function commit(S2) -> newS2`

Where both close over the device's endowments.  `write()` gets an `S` object that is an empty `Map` for the first call during the Crank, and a copy of the previous one for subsequent calls. Each `syscall.writeLater` causes an immediate device `write()` call, and `S` lets it accumulate both changes that need to be committed at the end of the crank, and a "shadow table" of data that helps figure out if the arguments are legal.

At the end of the Crank, the kernel invokes the `commit` function with a separate `S2` state object. `commit` should pull the queued writes from `S` and apply them to the endowment. It can also return a modified S2 object to retain non-endowment state from one Crank to the next. E.g. the Timer device would use S2 to record the set of callback handler objects for each timer, and the Mailbox would use it to record the inbound handler object.

**Comment by warner** [comment]
I added some notes to a new ticket before remembering this one, here are those notes:

The current swingset device model defines "devices" as containers very much like vats: the kernel dispatches some messages into them, they can make some syscalls back out through the kernel, there is a c-list between the two. Where vats export "objects", devices export "device nodes". But unlike vats, which are completely isolated from the outside world and can only communicate through the kernel, devices are given some collection of `endowments`: arbitrary javascript objects that can do whatever the host (who created them) likes.

Vats interact with each other by calling `syscall.send()`. The kernel translates the arguments through the vat's c-list, and puts the resulting message on the back of the kernel's run-queue. When the message gets to the front, it looks up the owner of the target object, translates the arguments through the target vat's c-list, and calls that vat's `dispatch.deliver()`. A similar pathway works between vats for Promises: one vat does `syscall.fulfillToData` (or one of its siblings), and some other vat(s) eventually get a `dispatch.notifyFulfillToData()`.

Vats interact with device nodes through a special syscall named `syscall.callNow()`. The kernel translates the arguments through the vat's c-list, figures out which device owns the device node, translates the arguments again through the *device*'s c-list, then dispatches into the device's `dispatch.invoke()`. The other special thing about devices is that vats can interact with them *synchronously*. Unlike `syscall.send()`, which returns (nothing) as soon as the message is queued, `syscall.callNow()` waits for the kernel to invoke the target device, and waits for that device to return a value. Whatever `dispatch.invoke()` returns is translated just like arguments would be, and is returned to the calling vat as the return value of `syscall.callNow()`. This is exposed to the upper-level code in a vat as the "D" invocation: `retval = D(devnode).methodname(args)`.

The idea was to provide a general-purpose (but capability-friendly) way for vats to interact with the outside world, mediated by devices. This makes `syscall.callNow()` roughly similar to a unix `ioctl()` call on a character device: arbitrary arguments, arbitrary return value, with synchronous/blocking semantics.

## read-old/write-later API

Dean has recommended a more structured API. Vats would be limited to invoking devices with either an explicit `read()` or an explicit `write()`. Both would accept arguments. The API would incorporate some kind of transaction boundary, perhaps a crank or a block. `read()` would be defined to return data about the state of the world at the beginning of the current temporal region: two `read()`s in the same block would always return the same data, no matter what `write()` calls appeared in between. And `write()` would be defined to queue up state changes that will be applied at the *end* of this temporal region. A read/write/read sequence would pretend that the write didn't happen. A write/write sequence, with arguments that cause the writes to overlap, would let the last one win.

## devices

Devices aren't just passive: the device itself gets to push messages onto the run-queue. We need this for any sort of inbound events:

* when the `mailbox` device receives new inbound messages, it does a `syscall.sendOnly` to the vat-tp/comms vat with the subset that were not already acked
* when the host informs the `timer` device about time passing, the device might trigger vat messages if an alarm/timeout occurs or a repeater has reached its next firing time
* inbound IBC messages get routed to a handler vat, and/or the vat-tp/comms vat

However these events only happen in the spaces between cranks/blocks: they do not (should not / must not) occur while a crank is running. So in a sense the inbound events are updating the "state of the world" before the block, and any device reads happening during the block should sample that updated state.
